### PR TITLE
proc: checks system security policy before trying to get personalities

### PR DESCRIPTION
**Pull request by HorlogeSkynet**
---

See #636.

### CODE REVIEW

**Comment by mihalicyn** [comment]
Dear @HorlogeSkynet,

please add Signed-off-by tag to your commit message.

File: src/utils.h [hunk 1]
```diff
@@ -25,6 +25,12 @@
 #define SEND_CREDS_NOTSK 1
 #define SEND_CREDS_FAIL 2
 
+#define SYS_FS_KERNEL_YAMA_PTRACE_SCOPE "/sys/kernel/yama/ptrace_scope"
+#define YAMA_PTRACE_SCOPE_CLASSIC 0
```
Comment (mihalicyn) [inline src/utils.h:29]: Let's align these mode names with what we have in Linux kernel https://github.com/torvalds/linux/blob/98369dccd2f8e16bf4c6621053af7aa4821dcf8e/security/yama/yama_lsm.c#L23

File: src/utils.c [hunk 2]
```diff
@@ -691,3 +691,42 @@ int get_task_personality(pid_t pid, __u32 *personality)
...
+	access mode check (see PROC(5)).
+*/
+bool is_ptrace_allowed(void)
+{
+	static int yama_ptrace_scope = -1;
+
+	__u32 buf_u32;
+	__do_close int fd = -EBADF;
+	int ret = -1;
+	char buf[8 + 1];
```
Comment (mihalicyn) [inline src/utils.c:707]: `char buf[INTTYPE_TO_STRLEN(uint32_t)];`

File: src/utils.c [hunk 3]
```diff
@@ -691,3 +691,42 @@ int get_task_personality(pid_t pid, __u32 *personality)
...
+		fd = open(SYS_FS_KERNEL_YAMA_PTRACE_SCOPE, O_RDONLY | O_CLOEXEC);
+		if (fd >= 0) {
+			ret = read_nointr(fd, buf, sizeof(buf) - 1);
+			if (ret >= 0) {
+				buf[ret] = '\0';
+				if (safe_uint32(buf, &buf_u32, 16) < 0)
+					return log_error(true, "Failed to read Yama ptrace scope %s", buf);
+
+				yama_ptrace_scope = buf_u32;
+			}
```
Comment (mihalicyn) [inline src/utils.c:723]: Let's print something in the logs in case when `read_nointr` fails.

**Review by mihalicyn — COMMENTED** [review]
LGTM, except a few small nits. Thanks!

File: src/proc_fuse.c [hunk 4]
```diff
@@ -145,8 +145,11 @@ __lxcfs_fuse_ops int proc_getattr(const char *path, struct stat *sb)
 	    strcmp(path, "/proc/swaps")		== 0 ||
 	    strcmp(path, "/proc/loadavg")	== 0 ||
 	    strcmp(path, "/proc/slabinfo")	== 0) {
-		if (liblxcfs_functional())
+		if (liblxcfs_functional()) {
+			if (!is_ptrace_allowed())
+				return log_error(-EACCES, RETRICTED_YAMA_PTRACE_POLICY);
```
Comment (stgraber) [inline src/proc_fuse.c:150]: Should probably be RESTRICTED rather than RETRICTED

Comment (stgraber) [reply src/proc_fuse.c:150]: Though that likely doesn't matter too much as we'll want to use the much simpler approach that's not yama-specific anyways :)

Comment (HorlogeSkynet) [reply src/proc_fuse.c]: Fixed, for posterity :upside_down_face:

**Pull request closed without merging**